regenerate-initrd-posttrans

```diff
@@ -15,14 +15,84 @@
 # to have all initrds generated. This script is called from posttrans
 # and takes care of generating the initrds
 
+# Check if `initrd_generator` is defined. See kernel-install(8) for details.
+# Note on how systemd gets this config:
+# - All the dropins are included in alphabetical order, no matter the location.
+# - If the variable is defined in any dropin, the main config file is ignored,
+# so in that case we don't need to find it.
+# - It only reads the first main config file (kernel/install.conf). From highest
+# to lowest priority: /etc, /usr/local/lib, /usr/lib.
+get_kernel_install_dropins_conf() {
+	kernel_install_dropins_conf=
+	for f in \
+		/etc/kernel/install.conf.d/*.conf \
+		/usr/local/lib/kernel/install.conf.d/*.conf \
+		/usr/lib/kernel/install.conf.d/*.conf \
+		; do
+		[ -f "$f" ] || continue
+		kernel_install_dropins_conf="$kernel_install_dropins_conf $f"
+	done
+	if [ -n "$kernel_install_dropins_conf" ]; then
+		for f in $kernel_install_dropins_conf; do
+			echo "${f##*/}"
+		done | sort -Vur | while read -r conf; do
+			for f in \
+				"/etc/kernel/install.conf.d/$conf" \
+				"/usr/local/lib/kernel/install.conf.d/$conf" \
+				"/usr/lib/kernel/install.conf.d/$conf" \
+				; do
+				[ -f "$f" ] || continue
+				. "$f"
+				if [ -n "$initrd_generator" ]; then
+					echo "$initrd_generator"
+					break 2
+				fi
+			done
+		done
+	fi
+}
+
+initrd_generator="$(get_kernel_install_dropins_conf)"
+if [ -z "$initrd_generator" ]; then
+	for f in \
+		/etc/kernel/install.conf \
+		/usr/local/lib/kernel/install.conf \
+		/usr/lib/kernel/install.conf \
+		; do
+		if [ -f "$f" ]; then
+			. "$f"
+			break
+		fi
+	done
+fi
+
+# dracut is the default initrd generator
+: ${initrd_generator:=dracut}
+
 : ${DRACUT:=/usr/bin/dracut}
-if [ ! -x "$DRACUT" ]; then
-    echo "${0##*/}: dracut is not installed, not rebuilding the initrd" >&2
-    exit 0
+if [ "$initrd_generator" = "dracut" ] && [ ! -x "$DRACUT" ]; then
+	echo "${0##*/}: dracut is not installed, not rebuilding the initrd" >&2
+	exit 0
+fi
+
+: ${MKOSI_INITRD:=/usr/bin/mkosi-initrd}
+[ "$initrd_generator" = "mkosi" ] && initrd_generator="mkosi-initrd"
```

Comment:
This will error out if the script is ever run with `bash -e`. Please use `||` or `if`.

Reply:
I suppose we *should* error out if `$initrd_generator` is neither `dracut` nor `mkosi`, but in that case we should print a meaningful error message.

Reply:
> This will error out if the script is ever run with `bash -e`. Please use `||` or `if`.

Actually, no. With `&&` (or `||`), it's a conditional expression, and failure of those expressions does not cause termination.

Reply:
I stand corrected. This is embarrassing. I could swear I'd observed the behavior I falsely described many times ... but maybe I've been fooled by something else, can't tell any more.

Thanks for pointing it out.

Reply:
:)